crates/core/tedge_agent/src/http_rest.rs

```diff
@@ -0,0 +1,299 @@
+use futures::StreamExt;
+use hyper::{server::conn::AddrIncoming, Request};
+use hyper::{Body, Response, Server};
+use routerify::{Router, RouterService};
+use std::{io::Write, path::Path};
+use std::{net::SocketAddr, path::PathBuf};
+
+#[cfg(not(test))]
+use tedge_config::{ConfigSettingAccessor, TmpPathSetting};
+use tedge_utils::paths::create_directories;
+use tokio::io::AsyncReadExt;
+
+use crate::error::AgentError;
+const FILE_TRANSFER_HTTP_ENDPOINT: &'static str = "/tedge/*";
+const WHAT_TO_NAME_THIS: &'static str = "tedge/file-transfer"; // TODO
+
+#[cfg(not(test))]
+lazy_static::lazy_static! {
+    static ref FILE_TRANSFER_ROOT_PATH: PathBuf = {
+        let config = tedge_config::get_tedge_config().unwrap();  // TODO
+        let value = config.query(TmpPathSetting).unwrap();  // TODO
+        let value: PathBuf = value.into();
+        let root_path = PathBuf::from(format!("{}/{WHAT_TO_NAME_THIS}", value.display()));
+        root_path
+    };
+}
+
+#[cfg(test)]
+lazy_static::lazy_static! {
+    static ref FILE_TRANSFER_ROOT_PATH: PathBuf = {
+        let ttd = tedge_test_utils::fs::TempTedgeDir::new();
+        dbg!(&ttd.path());
+        let root_path = PathBuf::from(format!("{}/{WHAT_TO_NAME_THIS}", ttd.path().display()));
+        root_path
+    };
+}
+
+fn separate_path_and_file_name(input: &str) -> Option<(PathBuf, &str)> {
+    let (mut relative_path, file_name) = input.rsplit_once("/")?;
+
+    if relative_path.starts_with("/") {
+        relative_path = &relative_path[1..];
+    }
+    let relative_path = PathBuf::from(relative_path);
+    Some((relative_path, file_name))
+}
+
+async fn put(mut request: Request<Body>) -> Result<Response<Body>, AgentError> {
+    let uri = request.uri().to_string();
+    let mut response = Response::new(Body::empty());
+    if let Some((relative_path, file_name)) = separate_path_and_file_name(&uri) {
+        let root_path = &FILE_TRANSFER_ROOT_PATH;
+        let directories_path = root_path.join(relative_path);
```

Comment:
This should be canonicalized.

Also, as far as I can see, there's not access control here. So something like `GET ../../../etc/passwd` would be possible, which is clearly a security incident here.

Reply:
I can only agree with @matthiasbeyer. I would even say, this must be canonicalized and check against the root path,